persist/sqldb/workflow_archive.go

```diff
@@ -30,81 +33,125 @@ type archivedWorkflowRecord struct {
-			res, err := r.session.
-				Update(archiveTableName).
-				Set("workflow", string(workflow)).
-				Set("phase", wf.Status.Phase).
-				Set("startedat", wf.Status.StartedAt.Time).
-				Set("finishedat", wf.Status.FinishedAt.Time).
-				Where(db.Cond{"clustername": r.clusterName}).
-				And(db.Cond{"uid": wf.UID}).
-				Exec()
-			if err != nil {
+	return r.session.Tx(context.Background(), func(sess sqlbuilder.Tx) error {
+		// We assume that we're much more likely to be inserting rows that updating them, so we try and insert,
+		// and if that fails, then we update.
+		// There is no check for race condition here, last writer wins.
+		_, err = sess.Collection(archiveTableName).
+			Insert(&archivedWorkflowRecord{
+				archivedWorkflowMetadata: archivedWorkflowMetadata{
+					ClusterName: r.clusterName,
+					UID:         string(wf.UID),
+					Name:        wf.Name,
+					Namespace:   wf.Namespace,
+					Phase:       wf.Status.Phase,
+					StartedAt:   wf.Status.StartedAt.Time,
+					FinishedAt:  wf.Status.FinishedAt.Time,
+				},
+				Workflow: string(workflow),
+			})
+		if err != nil {
+			if isDuplicateKeyError(err) {
+				res, err := sess.
+					Update(archiveTableName).
+					Set("workflow", string(workflow)).
+					Set("phase", wf.Status.Phase).
+					Set("startedat", wf.Status.StartedAt.Time).
+					Set("finishedat", wf.Status.FinishedAt.Time).
+					Where(db.Cond{"clustername": r.clusterName}).
+					And(db.Cond{"uid": wf.UID}).
+					Exec()
+				if err != nil {
+					return err
+				}
+				rowsAffected, err := res.RowsAffected()
+				if err != nil {
+					return err
+				}
+				if rowsAffected != 1 {
+					logCtx.WithField("rowsAffected", rowsAffected).Warn("Expected exactly one row affected")
+				}
+			} else {
 				return err
 			}
-			rowsAffected, err := res.RowsAffected()
+		}
+
+		// insert the labels
+		for key, value := range wf.GetLabels() {
+			_, err := sess.Collection(archiveLabelsTableName).
+				Insert(&archivedWorkflowLabelRecord{
+					ClusterName: r.clusterName,
+					Uid:         string(wf.UID),
+					Key:         key,
+					Value:       value,
+				})
 			if err != nil {
-				return err
+				if isDuplicateKeyError(err) {
+					_, err = sess.
+						Update(archiveLabelsTableName).
+						Set("value", value).
+						Where(db.Cond{"clustername": r.clusterName}).
+						And(db.Cond{"uid": wf.UID}).
+						And(db.Cond{"name": key}).
+						Exec()
+					if err != nil {
+						return err
+					}
+				} else {
+					return err
+				}
 			}
-			if rowsAffected != 1 {
-				logCtx.WithField("rowsAffected", rowsAffected).Warn("Expected exactly one row affected")
-			}
-		} else {
-			return err
-		}
-	}
 
-	return nil
+		}
+		return nil
+	})
 }
 
-func (r *workflowArchive) ListWorkflows(namespace string, limit int, offset int) (wfv1.Workflows, error) {
+func (r *workflowArchive) ListWorkflows(namespace string, labelRequirements labels.Requirements, limit int, offset int) (wfv1.Workflows, error) {
```

Comment:
Labels key-values need to be set in the returned workflow list.

Reply:
why?

Reply:
Those workflow labels will be used for label suggestions in the filter component, considering it's already paginated, having them shouldn't bring much performance concern.

Reply:
I'm not sure how we'd do this, joining onto argo_archived_workflow_label would be a cartesian join, and selecting from the workflow column would negatively impact performance. 

Unless this is a must-have - I don't think we can do it.